### ISSUE TITLE
Documents with duplicate ids are being added normally

### ISSUE THREAD

**Issue by lilsweetcaligula** [lucaong/minisearch#101]
Hello. Thank you for your hard work on this package!

I think I may have encountered a problem with the ways ids are treated. In particular, when adding two documents with the same id, both documents are added to the search pool. No error is raised, no upsert is being carried out - both documents with the same id are added normally as if they had different ids. I am not sure if that's expected behavior.

The following example code:
```javascript
const Minisearch = require('minisearch')


async function run() {
  const minisearch = new Minisearch({
    fields: ['value']
  })

  minisearch.add({ id: 'b', value: 'bob' })
  minisearch.add({ id: 'b', value: 'boba' })


  const ans = minisearch.search('bob', { fuzzy: true })

  console.dir(ans, { depth: 4 })
}


run()
```

outputs:
```javascript
[
  {
    id: 'b',
    terms: [ 'bob' ],
    score: 2.0794415416798357,
    match: { bob: [ 'value' ] }
  },
  {
    id: 'b',
    terms: [ 'boba' ],
    score: 0.6398773880082578,
    match: { boba: [ 'value' ] }
  }
]
```

For convenience, I have created a [repository](https://github.com/lilsweetcaligula/minisearch-ids-bug-report) which reproduces the issue outlined in the example above.

Kind Regards,
lilsweetcaligula

**Comment by lucaong** [comment]
Hi @lilsweetcaligula ,
Thanks for reporting this. I agree that it would be better to throw an error if a document is added having an id that is already in the index. It might be a bit tricky to implement without additionally saving a map of all ids, or performing a preliminary search upon adding a document, which would make the memory utilization or the insert performance worse.

Let me think if there is a simple way to fix this.

**Comment by rossbu** [comment]
+1 , good point. may need to raise a framework solution for exception handling.

**Comment by lucaong** [comment]
For those who need this, the check for duplicates is relatively easy to implement in user code. One could, in the simplest case, maintain a set of all document IDs, and check for existence when adding one:

```javascript
const miniSearch = new MiniSearch(...)
const ids = new Set()

const addDocument = function (document) {
  if (ids.has(document.id)) {
    throw new Error(`Duplicated document ID ${document.id}`)
  }
  ids. add(document.id)
  miniSearch.add(document)
}
```

As a matter of fact, in most cases one would wrap this in a wrapper class, so it can also include more logic. Here for example the wrapper keeps a map of documents to return from a search directly matching documents as opposed to search results with IDs:

```javascript
class SearchEngine {
  constructor (options) {
    this.miniSearch = new MiniSearch(options)
    this.documentsById = {}
  }

  addOne (document) {
    if (this.documentsById[document.id] != null) {
      throw new Error(`Duplicated document ID ${document.id}`)
    }
    this.documentsById[document.id] = document
    this.miniSearch.add(document)
  }

  addAll (documents) {
    documents.forEach((document) => {
      this.addOne(document)
    })
  }

  search (query, options = {}) {
    return this.miniSearch.search(query, options).map((result) => this.documentsById[result.id])
  }
}
```

The examples above assume that the ID field is `id`, but relaxing this limitation is possible with a little more code.

**Comment by KeKs0r** [comment]
I also just came across this issue. Unfortunately just ignoring duplicates would not work, since the reason I am trying to readd the same document is, that I know the document itself changed. 
My callback is also triggered with all documents so just removing a single and readding is not possible. 

I know that is more a userland problem, and I need to better identify the actual change that is happening, but thought I mention it of being the case.

**Comment by lucaong** [comment]
I have some changes in mind that would help with this issue, as well as making it possible to delete a document by ID, and even “patch” a document by re-indexing or adding only specific fields (at least when using some specific configuration), but they involve making the serialization format not backward compatible, so they will need a major release. I also need to check the implications on memory utilization.

**Comment by lucaong** [comment]
> I have some changes in mind that would help with this issue, as well as making it possible to delete a document by ID, and even “patch” a document by re-indexing or adding only specific fields (at least when using some specific configuration), but they involve making the serialization format not backward compatible, so they will need a major release. I also need to check the implications on memory utilization.

I do not have a timeline for this feature though.

**Comment by lucaong** [comment]
@lilsweetcaligula @rafaellichen @rossbu @KeKs0r I revive this thread to announce that the newly released `v6.0.0-beta.1` (soon to be `v6.0.0`) includes the feature you mention:

  - The methods `add` and `addAll` now throw an error on duplicate ID
  - A new method `has(id)` can be used to check for the existence of a document by ID
  - A new method `discard` (and `discardAll`) can be used to remove a document by ID
  - A new method `replace` can be used to replace a document with a new version of it

See [the documentation](https://lucaong.github.io/minisearch/classes/_minisearch_.minisearch.html) for more information.
